File: src/types/agents.ts
export interface AgentValidationDiagnostic {
  severity: string;
  field: string | null;
  message: string;
}

export interface AgentDefinition {
  agent_id: string;
  agent_name: string;
  agent_version: string;
  author: string;
  description: string;
  tags: string[];
  path: string;
  avatar_path: string;
  readme_path: string;
  bundled: boolean;
  enabled: boolean;
  lifecycle_state: string;
  rag_enabled: boolean;
  is_multi_agent_supported: boolean;
  has_rag_knowledge: boolean;
  validation_diagnostics: AgentValidationDiagnostic[];
}

export interface AgentManifest {
  agent_id: string;
  agent_name: string;
  agent_version: string;
  author: string;
  description: string;
  tags: string[];
  create_time: string;
  update_time: string;
  min_app_version: string;
  rag_enabled: boolean;
  is_multi_agent_supported: boolean;
}

export interface AgentConfig {
  temperature: number;
  top_p: number;
  rag_top_k: number;
  embedding_dim: number;
  response_style: string;
  ban_topics: string[];
}

export interface AgentDefinitionDetail {
  agent: AgentDefinition;
  manifest: AgentManifest | null;
  config: AgentConfig | null;
  system_prompt: string;
  readme: string;
}

export interface InstallAgentZipInput {
  zip_path: string;
}

export interface AgentDirectorySettings {
  agent_directory: string;
}

export interface SaveAgentDirectorySettings {
  agent_directory: string;
}

export interface AgentSafeFileRootSettings {
  safe_file_roots: string[];
}

export interface SaveAgentSafeFileRootSettings {
  safe_file_roots: string[];
}

export interface AgentMessage {
  message_id: string;
  session_id: string;
  sender_type: number;
  agent_id: string | null;
  content: string;
  turn_index: number;
  stream_status: string;
  error_text: string;
  created_at: string;
}

export interface AgentSession {
  session_id: string;
  session_type: number;
  agent_ids: string[];
  session_title: string;
  memory_enabled: boolean;
  messages: AgentMessage[];
  created_at: string;
  updated_at: string;
}

export interface SendAgentMessageResult {
  session: AgentSession;
  assistant_message: AgentMessage;
  used_context: AgentUsedContext;
}

export interface AgentUserIdentity {
  display_name: string;
  preferred_language: string;
  communication_style: string;
  roles: string[];
  goals: string[];
  boundaries: string;
  important_facts: string;
  enabled: boolean;
  updated_at: string;
}

export interface SaveAgentUserIdentity {
  display_name: string;
  preferred_language: string;
  communication_style: string;
  roles: string[];
  goals: string[];
  boundaries: string;
  important_facts: string;
  enabled: boolean;
}

export interface AgentMemory {
  memory_id: string;
  content: string;
  scope: string;
  agent_id: string | null;
  status: string;
  pinned: boolean;
  source_session_id: string | null;
  source_agent_id: string | null;
  source_message_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface AgentMemoryProposal {
  proposal_id: string;
  source_session_id: string | null;
  source_agent_id: string | null;
  source_message_id: string | null;
  proposed_text: string;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface SaveAgentMemory {
  memory_id?: string | null;
  content: string;
  scope: string;
  agent_id?: string | null;
  status?: string | null;
  pinned?: boolean | null;
  source_session_id?: string | null;
  source_agent_id?: string | null;
  source_message_id?: string | null;
}

export interface ConfirmAgentMemoryProposalInput {
  proposal_id: string;
  accepted: boolean;
  content?: string | null;
  scope?: string | null;
  agent_id?: string | null;
}

export interface AgentUsedContext {
  todos: number[];
  milestones: number[];
  notes: number[];
  memories: string[];
  rag_chunks: string[];
  conversation_summaries: string[];
  previous_messages: string[];
}

export interface AgentStreamEvent {
  stream_id: string;
  agent_id?: string | null;
  content?: string;
  error?: string;
  result?: SendAgentMessageResult;
}

export interface AgentRagStatus {
  agent_id: string;
  rag_enabled: boolean;
  has_rag_knowledge: boolean;
  indexed_chunks: number;
  source_hash: string | null;
  current_source_hash: string | null;
  stale: boolean;
  stale_reasons: string[];
  vector_search_available: boolean;
  message: string;
}

export interface AgentMigrationStatus {
  migration_id: string;
  status: string;
  details: string;
  created_at: string;
  updated_at: string;
}

export interface AgentBuiltinTool {
  name: string;
  description: string;
  permission_category: string;
  safety_class: string;
  requires_confirmation: boolean;
  argument_schema: unknown;
  result_schema: unknown;
}

export interface AgentExternalCliTool {
  tool_id: string;
  display_name: string;
  executable: string;
  allowed_subcommands: string[];
  argument_schema: unknown;
  working_directory: string;
  environment_allowlist: string[];
  timeout_ms: number;
  output_limit: number;
  safety_class: string;
  enabled: boolean;
  available: boolean;
  availability_error: string;
  created_at: string;
  updated_at: string;
}

export interface SaveAgentExternalCliTool {
  tool_id?: string | null;
  display_name: string;
  executable: string;
  allowed_subcommands: string[];
  argument_schema: unknown;
  working_directory: string;
  environment_allowlist: string[];
  timeout_ms: number;
  output_limit: number;
  safety_class: string;
  enabled: boolean;
}

export interface AgentExternalCliToolTestResult {
  executable: string;
  resolved_path: string;
  available: boolean;
  message: string;
}

export interface AgentToolAction {
  action_id: string;
  session_id: string | null;
  agent_id: string | null;
  tool_name: string;
  arguments: unknown;
  preview: unknown;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface ConfirmAgentToolActionResult {
  action_id: string;
  accepted: boolean;
  tool_name: string;
  status: string;
  result: unknown;
}

export type { SelectedAppContext };
import type { SelectedAppContext } from "./secretary";

File: src-tauri/src/models/agents.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::models::secretary::SelectedAppContext;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentManifest {
    #[serde(alias = "plugin_id")]
    pub agent_id: String,
    #[serde(alias = "plugin_name")]
    pub agent_name: String,
    #[serde(alias = "plugin_version")]
    pub agent_version: String,
    pub author: String,
    pub description: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub create_time: String,
    pub update_time: String,
    pub min_app_version: String,
    pub rag_enabled: bool,
    pub is_multi_agent_supported: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentConfig {
    pub temperature: f64,
    pub top_p: f64,
    pub rag_top_k: i64,
    pub embedding_dim: i64,
    pub response_style: String,
    #[serde(default)]
    pub ban_topics: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentValidationDiagnostic {
    pub severity: String,
    pub field: Option<String>,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentDefinition {
    pub agent_id: String,
    pub agent_name: String,
    pub agent_version: String,
    pub author: String,
    pub description: String,
    pub tags: Vec<String>,
    pub path: String,
    pub avatar_path: String,
    pub readme_path: String,
    pub bundled: bool,
    pub enabled: bool,
    pub lifecycle_state: String,
    pub rag_enabled: bool,
    pub is_multi_agent_supported: bool,
    pub has_rag_knowledge: bool,
    pub validation_diagnostics: Vec<AgentValidationDiagnostic>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentDefinitionDetail {
    pub agent: AgentDefinition,
    pub manifest: Option<AgentManifest>,
    pub config: Option<AgentConfig>,
    pub system_prompt: String,
    pub readme: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentDirectorySettings {
    pub agent_directory: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaveAgentDirectorySettings {
    pub agent_directory: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstallAgentZipInput {
    pub zip_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentSafeFileRootSettings {
    pub safe_file_roots: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaveAgentSafeFileRootSettings {
    #[serde(default)]
    pub safe_file_roots: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRagChunk {
    pub chunk_id: String,
    pub agent_id: String,
    pub agent_version: String,
    pub source_hash: String,
    pub embedding_model: String,
    pub embedding_dim: i64,
    pub chunk_text: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRagStatus {
    pub agent_id: String,
    pub rag_enabled: bool,
    pub has_rag_knowledge: bool,
    pub indexed_chunks: usize,
    pub source_hash: Option<String>,
    pub current_source_hash: Option<String>,
    pub stale: bool,
    #[serde(default)]
    pub stale_reasons: Vec<String>,
    pub vector_search_available: bool,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMigrationStatus {
    pub migration_id: String,
    pub status: String,
    pub details: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMessage {
    pub message_id: String,
    pub session_id: String,
    pub sender_type: i64,
    pub agent_id: Option<String>,
    pub content: String,
    pub turn_index: i64,
    pub stream_status: String,
    pub error_text: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentSession {
    pub session_id: String,
    pub session_type: i64,
    pub agent_ids: Vec<String>,
    pub session_title: String,
    pub memory_enabled: bool,
    pub messages: Vec<AgentMessage>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentUserIdentity {
    pub display_name: String,
    pub preferred_language: String,
    pub communication_style: String,
    pub roles: Vec<String>,
    pub goals: Vec<String>,
    pub boundaries: String,
    pub important_facts: String,
    pub enabled: bool,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaveAgentUserIdentity {
    pub display_name: String,
    pub preferred_language: String,
    pub communication_style: String,
    #[serde(default)]
    pub roles: Vec<String>,
    #[serde(default)]
    pub goals: Vec<String>,
    pub boundaries: String,
    pub important_facts: String,
    pub enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMemory {
    pub memory_id: String,
    pub content: String,
    pub scope: String,
    pub agent_id: Option<String>,
    pub status: String,
    pub pinned: bool,
    pub source_session_id: Option<String>,
    pub source_agent_id: Option<String>,
    pub source_message_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentMemoryProposal {
    pub proposal_id: String,
    pub source_session_id: Option<String>,
    pub source_agent_id: Option<String>,
    pub source_message_id: Option<String>,
    pub proposed_text: String,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentConversationSummary {
    pub summary_id: String,
    pub session_id: String,
    pub agent_id: Option<String>,
    pub title: String,
    pub summary: String,
    pub topics: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaveAgentMemory {
    pub memory_id: Option<String>,
    pub content: String,
    pub scope: String,
    pub agent_id: Option<String>,
    pub status: Option<String>,
    pub pinned: Option<bool>,
    pub source_session_id: Option<String>,
    pub source_agent_id: Option<String>,
    pub source_message_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmAgentMemoryProposalInput {
    pub proposal_id: String,
    pub accepted: bool,
    pub content: Option<String>,
    pub scope: Option<String>,
    pub agent_id: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AgentUsedContext {
    #[serde(default)]
    pub todos: Vec<i64>,
    #[serde(default)]
    pub milestones: Vec<usize>,
    #[serde(default)]
    pub notes: Vec<i64>,
    #[serde(default)]
    pub memories: Vec<String>,
    #[serde(default)]
    pub rag_chunks: Vec<String>,
    #[serde(default)]
    pub conversation_summaries: Vec<String>,
    #[serde(default)]
    pub previous_messages: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendAgentMessageInput {
    pub session_id: Option<String>,
    pub agent_id: String,
    pub message: String,
    #[serde(default)]
    pub selected_context: SelectedAppContext,
    pub stream_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendAgentGroupMessageInput {
    pub session_id: Option<String>,
    #[serde(default)]
    pub agent_ids: Vec<String>,
    pub message: String,
    #[serde(default)]
    pub selected_context: SelectedAppContext,
    pub stream_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendAgentMessageResult {
    pub session: AgentSession,
    pub assistant_message: AgentMessage,
    pub used_context: AgentUsedContext,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentBuiltinTool {
    pub name: String,
    pub description: String,
    pub permission_category: String,
    pub safety_class: String,
    pub requires_confirmation: bool,
    pub argument_schema: Value,
    pub result_schema: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentToolCallInput {
    pub session_id: Option<String>,
    pub agent_id: Option<String>,
    pub tool_name: String,
    #[serde(default)]
    pub arguments: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentToolCallResult {
    pub audit_id: String,
    pub action_id: Option<String>,
    pub tool_name: String,
    pub status: String,
    pub requires_confirmation: bool,
    pub result: Value,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentToolAction {
    pub action_id: String,
    pub session_id: Option<String>,
    pub agent_id: Option<String>,
    pub tool_name: String,
    pub arguments: Value,
    pub preview: Value,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmAgentToolActionInput {
    pub action_id: String,
    pub accepted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmAgentToolActionResult {
    pub action_id: String,
    pub accepted: bool,
    pub tool_name: String,
    pub status: String,
    pub result: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentExternalCliTool {
    pub tool_id: String,
    pub display_name: String,
    pub executable: String,
    #[serde(default)]
    pub allowed_subcommands: Vec<String>,
    #[serde(default)]
    pub argument_schema: Value,
    pub working_directory: String,
    #[serde(default)]
    pub environment_allowlist: Vec<String>,
    pub timeout_ms: i64,
    pub output_limit: i64,
    pub safety_class: String,
    pub enabled: bool,
    pub available: bool,
    pub availability_error: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaveAgentExternalCliTool {
    pub tool_id: Option<String>,
    pub display_name: String,
    pub executable: String,
    #[serde(default)]
    pub allowed_subcommands: Vec<String>,
    #[serde(default)]
    pub argument_schema: Value,
    pub working_directory: String,
    #[serde(default)]
    pub environment_allowlist: Vec<String>,
    pub timeout_ms: i64,
    pub output_limit: i64,
    pub safety_class: String,
    pub enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentExternalCliToolTestResult {
    pub executable: String,
    pub resolved_path: String,
    pub available: bool,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentExternalCliCallInput {
    pub session_id: Option<String>,
    pub agent_id: Option<String>,
    pub tool_id: String,
    #[serde(default)]
    pub arguments: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentExternalCliCallResult {
    pub audit_id: String,
    pub tool_id: String,
    pub status: String,
    pub confirmation_status: String,
    pub exit_code: Option<i32>,
    pub stdout: String,
    pub stderr: String,
    pub duration_ms: i64,
    pub timed_out: bool,
    pub truncated: bool,
    pub message: String,
}
